File: benchmarks/covariance/covariance.jl
using Statistics
using LinearAlgebra
using PrettyTables
using DelimitedFiles

include("../../timing/dphpc_timing.jl")
DEV = true
TIME = true

eval_benchmarks = Dict(
    "S" => (500, 600),
    "M" => (1400, 1800),
    "L" => (3200, 4000),
    "paper" => (1200, 1400)
)

dev_benchmarks = Dict(
    "S" => (2, 2),
    "M" => (5, 7),
    "L" => (10, 20),
)

function main() 
    if DEV benchmark_sizes = dev_benchmarks 
    else benchmark_sizes = eval_benchmarks end

    for dims in keys(benchmark_sizes)
        println("Benchmarking $dims")
        M, N = benchmark_sizes[dims]

        data = initialize(M, N)

        if TIME
            res = @dphpc_time data, kernel(M, N, data)
            println(res)
        else
            res = kernel(M, N, data)
            pretty_table(res)
        end

    end

end

function kernel(M, float_n, data)
    mean_data = mean(data, dims=1)
    data .-= mean_data
    cov = zeros(eltype(data), M, M)
    for i in 1:M
        for j in i:M
            cov[i, j] = cov[j, i] = dot(data[:, i], data[:, j]) / (float_n - 1.0)
        end
    end
    return cov
end

function initialize(M, N, datatype=Float64)
    return [datatype((i-1) * (j-1)) / M for i in 1:N, j in 1:M]
end

main()

File: benchmarks/covariance/covariance.py
import numpy as np
import os

input_m = 2
input_n = 2

dev_benchmarks = {
    # "S": (2, 2),
    "M": (5, 5),
    "L": (5, 7),
    # "L": (10, 20),
}

def main():
    for benchmark in dev_benchmarks:
        M, N = dev_benchmarks[benchmark]
        print(f"Running on set M={M}, N={N}")

        data = initialize(M, N, datatype=np.float64)
        print("Data")
        print(data)
        print("Cov")
        cov = kernel(M, np.float64(N), data)
        print(cov)
        print("End")


def initialize(M, N, datatype=np.float64):
    return np.fromfunction(lambda i, j: (i * j) / M, (N, M), dtype=datatype)

def kernel(M, float_n, data):
    mean = np.mean(data, axis=0)
    print("mean", mean)
    data -= mean
    cov = np.zeros((M, M), dtype=data.dtype)
    for i in range(M):
        cov[i:M, i] = cov[i, i:M] = data[:, i] @ data[:, i:M] / (float_n - 1.0)

    return cov


if __name__ == "__main__":
    main()